src/renderers/shaders/ShaderChunk/tonemapping_pars_fragment.glsl.js

```diff
@@ -165,5 +165,28 @@ vec3 AgXToneMapping( vec3 color ) {
 
 }
 
+// https://modelviewer.dev/examples/tone-mapping
+
+vec3 CommerceToneMapping( vec3 color ) {
+	float startCompression = 0.8 - 0.04;
+	float desaturation = 0.15;
+
+	color *= toneMappingExposure;
+
+	float x = min(color.r, min(color.g, color.b));
+	float offset = x < 0.08 ? x - 6.25 * x * x : 0.04;
+	color -= offset;
+
+	float peak = max(color.r, max(color.g, color.b));
+	if (peak < startCompression) return color;
```

Comment:
@sunag I've experimented a bit with TSL and tried to implement this function by myself 😇 . Current code:

```js
const CommerceToneMappingNode = tslFn( ( { color, exposure } ) => {

	const startCompression = float( 0.8 - 0.04 );
	const desaturation = float( 0.15 );

	color.mulAssign( exposure );

	const x = float( min( color.r, min( color.g, color.b ) ) );
	const offset = float( cond( x.lessThan( 0.08 ), x.sub( mul( 6.25, x.mul( x ) ) ), 0.04 ) );

	color.subAssign( offset );

	const peak = float( max( color.r, max( color.g, color.b ) ) );

	If( peak.lessThan( startCompression ), () => {

		return color;

	} );

	const d = float( sub( 1, startCompression ) );
	const newPeak = float( sub( 1, d.mul( d ).div( peak.add( d.sub( startCompression ) ) ) ) );
	color.mulAssign( newPeak.div( peak ) );
	const g = float( sub( 1, div( 1, desaturation.mul( peak.sub( newPeak ) ).add( 1 ) ) ) );

	return mix( color, vec3( 1, 1, 1 ), g );

} );
```
After comparing with the TSL transpiler, I wonder where to import the `If` syntax? The transpiler imports it from `three/nodes`. But what is the concrete node class for importing `If`? I couldn't find it so far.

Reply:
`If` isn't a node, it is export from the `ShaderNode.js`.
Also, you can use some TSL-specific syntax -- for example, you can write `oneMinus` and `reciprocal` instead of `sub( 1, ... )` and `div( 1, ... )`. You also don't need any `float` type casters (they are only used for explicitly converting one type to another, which isn't needed here). I also suggest you to use some more node chaining (this reduces the need to import anything apart from `tslFn` and `vec3` here). Also, it's better to not use any assigns when not strictly needed -- they doesn't allow the Nodes system to do as much symplification.
I would simplify this code to something like:
```js
const CommerceToneMappingNode = tslFn( ( { color, exposure } ) => {

 	const startCompression = 0.8 - 0.04;
 	const desaturation = 0.15; 

	color = color.mul( exposure );

 	const x = color.r.min( color.g, color.b );
 	const offset = x.lessThan( 0.08 ).cond( x.sub( x.mul( x, 6.25 ) ), 0.04 );

 	color = color.sub( offset );

 	const peak = color.r.max( color.g, color.b );

 	const d = startCompression.oneMinus();
 	const newPeak = d.mul( d ).div( peak.add( d.sub( startCompression ) ) ).oneMinus();

 	color = color.mul( newPeak.div( peak ) );

 	const g = desaturation.mul( peak.sub( newPeak ) ).add( 1 ).reciprocal().oneMinus();

 	return peak.lessThan( startCompression ).cond( color, g.mix( color, vec3( 1, 1, 1 ) ) );

 } );
```
Note that the last `cond` **currently** will generate previous nodes (`d`, `newPeak`, `g`) in any case, even if they aren't use in a clause -- you can temporary revert to using `If` so that wouldn't happen, but we plan to fix it in any case.

Reply:
I think I would prefer to keep using `If` since that expresses the developer intent more clearly (meaning to perform an early out). That also makes the code more readable.

Removing the `float()` calls is indeed a good thing. However, I would also prefer to retain the original code structure for better readability and not chain too much.

Reply:
> After comparing with the TSL transpiler, I wonder where to import the If syntax? The transpiler imports it from three/nodes. But what is the concrete node class for importing If? I couldn't find it so far.

@Mugen87 Defining a return through an `if` conditional should work soon. ( Related https://github.com/mrdoob/three.js/pull/27542#issuecomment-1888357025 )

Reply:
We can add the node version of the new tone mapping at a later point as well so this PR is not blocked 👍 .